### PR TITLE
ci: build TSan with clang 15 and add -Werror=thread-safety, fix-up stacktraces

### PR DESCRIPTION
**Pull request by kittywhiskers**
## Description

Pull request was inspired by the need to debug lock problems when working on https://github.com/dashpay/dash/pull/5352.

As far as I'm aware, only macOS has `-Werror=thread-safety` as part of its default `CXXFLAGS` despite the capability being present on Linux as well. This PR introduces thread safety checks for that into our thread sanitizer build.

Additionally, since we're using Clang, something that on first glimpse, appears to be something that `stacktraces.cpp` isn't happy with, due to `-Wl,-wrap` being available only on GCC, that no longer seems to be the case, since the version of Clang with comes with `focal`, its `lld` _does_ have support for `-wrap` (see [man page for `lld` on `focal`](https://manpages.ubuntu.com/manpages/focal/en/man1/lld.1.html)).

The current `stable` version of Clang/LLVM is 15, at the time of this pull request (see https://apt.llvm.org/) but `focal` ships with an older version, requiring us to use the official LLVM APT repository. I feel we should be testing with recent compilers alongside the ones shipped by LTS distributions. 

Certain bugs are only made apparent when testing on rolling release distros or distros that have faster update cycles, like Fedora (see https://github.com/dashpay/dash/pull/5295 for an illustration of that), which ship with more recent compilers. Until we overhaul our CI systems to test using those distros directly (our current infrastructure is centered around using a "development image" with an LTS distro as the base), this is the best we can do.

A similar pull request testing against the latest GCC stable will be welcome as that is currently outside the scope of this PR as the changes made were to make sure that builds were operating as expected on Clang/LLVM 15.

### CODE REVIEW

File: contrib/containers/ci/Dockerfile [hunk 1]
```diff
@@ -77,8 +77,7 @@ RUN apt-get update && apt-get install $APT_ARGS \
     wine-stable \
     wine32 \
     wine64 \
-    xorriso \
-    && rm -rf /var/lib/apt/lists/*
```
Comment (PastaPastaPasta) [inline contrib/containers/ci/Dockerfile:81]: why change this?

File: src/stacktraces.cpp [hunk 2]
```diff
@@ -536,12 +536,9 @@ static std::map<void*, std::shared_ptr<std::vector<uint64_t>>> g_stacktraces;
 
 #if CRASH_HOOKS_WRAPPED_CXX_ABI
 // These come in through -Wl,-wrap
-// It only works on GCC
 extern "C" void* __real___cxa_allocate_exception(size_t thrown_size);
 extern "C" void __real___cxa_free_exception(void * thrown_exception);
-#if __clang__
-#error not supported on WIN32 (no dlsym support)
-#elif WIN32
+#if WIN32
```
Comment (PastaPastaPasta) [inline src/stacktraces.cpp]: also, feels like these changes should be in another PR? This PR should just be adding clang-15 and thread-safety

File: contrib/containers/ci/Dockerfile (hunk 1 above)
Comment (kittywhiskers) [reply contrib/containers/ci/Dockerfile:81]: Based on what I'm familiar with, running `rm -rf /var/lib/apt/lists/*` is done to reduce the size of the layer but Dash's Dockerfiles only runs it on the last invocation of `apt-get` instead of after _every_ invocation (see an invocation without the removal [here](https://github.com/dashpay/dash/blob/1d6990a/contrib/containers/ci/Dockerfile#L33)). 

Keeping that in mind, since this is no longer the last _manual_ invocation, the call is moved [here](https://github.com/dashpay/dash/blame/008461341fc7d2d54acd0aaf3578089539428f7a/contrib/containers/ci/Dockerfile#L110).

It seems like it would make more sense to add that line into every invocation but that's outside the scope of this PR.

File: src/stacktraces.cpp (hunk 2 above)
Comment (kittywhiskers) [reply src/stacktraces.cpp]: The fixes for stacktraces are directly aimed to resolve problems with a Clang/LLVM target, which didn't exist in our CI suite until this PR, so it was retained here. The changes weren't significant enough to warrant its own PR and aren't verifiable without this PR's merger (sans stracktrace changes)

**Review by UdjinM6 — APPROVED** [review]
LGTM, utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for squash merge
